lib/runtime-core/src/parse.rs

```diff
@@ -266,28 +266,22 @@ pub fn read_module<
                                 body_begun = true;
                                 fcg.begin_body(&info.read().unwrap())
                                     .map_err(|x| LoadError::Codegen(format!("{:?}", x)))?;
-                                middlewares
-                                    .run(
-                                        Some(fcg),
-                                        Event::Internal(InternalEvent::FunctionBegin(id as u32)),
-                                        &info.read().unwrap(),
-                                    )
-                                    .map_err(|x| LoadError::Codegen(x))?;
+                                events
+                                    .push(Event::Internal(InternalEvent::FunctionBegin(id as u32)));
                             }
-                            middlewares
-                                .run(Some(fcg), Event::Wasm(op), &info.read().unwrap())
-                                .map_err(|x| LoadError::Codegen(x))?;
+                            events.push(Event::WasmOwned(op.clone()));
+                        }
+                        ParserState::EndFunctionBody => {
+                            events.push(Event::Internal(InternalEvent::FunctionEnd));
+                            break;
                         }
-                        ParserState::EndFunctionBody => break,
                         _ => unreachable!(),
                     }
                 }
+
+                let events = EventSink { buffer: events };
```

Comment:
Collecting all events into one big sink here would increase memory usage, especially for large functions. Is this really needed?